### PR TITLE
Add cuda-cupti recipe

### PR DESCRIPTION
**Pull request by adibbley**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by jakirkham** [comment]
cc @bdice

File: recipes/cuda-cupti/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+outputs:
+  - name: cuda-cupti
+    test:
+      commands:
+        - test -f $PREFIX/lib/libcupti.so.*                # [linux]
+        - if not exist %LIBRARY_BIN%\cupti64_*.dll exit 1  # [win]
+    requirements:
+      run:
+        - cuda-version =12.0
```
Comment (bdice) [inline recipes/cuda-cupti/meta.yaml:38]:
```suggestion
        - cuda-version {{ cuda_version }}
```

File: recipes/cuda-cupti/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+      commands:
+        - test -f $PREFIX/lib/libnvperf_host.so        # [linux]
+        - if not exist %LIBRARY_LIB%\cupti.lib exit 1  # [win]
+    requirements:
+      build:
+        - {{ compiler("c") }}
+        - {{ compiler("cxx") }}
+      run:
+        - cuda-cupti >={{ version }}
+        - cuda-version =12.0
```
Comment (bdice) [inline recipes/cuda-cupti/meta.yaml:55]:
```suggestion
        - cuda-version {{ cuda_version }}
```

File: recipes/cuda-cupti/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+  - name: cuda-cupti-static
+    build:
+      skip: True                        # [not linux]
+    test:
+      commands:
+        - test -L $PREFIX/lib/libcupti_static.a  # [linux]
+    requirements:
+      run:
+        - cuda-cupti >={{ version }}
+        - cuda-version =12.0
```
Comment (bdice) [inline recipes/cuda-cupti/meta.yaml:75]:
```suggestion
        - cuda-version {{ cuda_version }}
```

**Comment by jakirkham** [comment]
Thanks Bradley! 🙏

These are good suggestions.

For `cuda-version` do we want to pin like this or use `{{ pin_compatible("cuda-version") }}` as suggested in `libcublas`? Think whatever we decide to that question it would be good to do consistently across CTK packages

File: recipes/cuda-cupti/meta.yaml (hunk 1 above)
Comment (bdice) [inline recipes/cuda-cupti/meta.yaml:38]:
```suggestion
    requirements:
      host:
        - cuda-version {{ cuda_version}}
      run:
        - {{ pin_compatible("cuda-version", max_pin="x.x") }}
```

File: recipes/cuda-cupti/meta.yaml (hunk 2 above)
Comment (bdice) [inline recipes/cuda-cupti/meta.yaml:55]:
```suggestion
    requirements:
      build:
        - {{ compiler("c") }}
        - {{ compiler("cxx") }}
      host:
        - cuda-version {{ cuda_version }}
      run:
        - {{ pin_compatible("cuda-version", max_pin="x.x") }}
        - cuda-cupti >={{ version }}
```

File: recipes/cuda-cupti/meta.yaml (hunk 3 above)
Comment (bdice) [inline recipes/cuda-cupti/meta.yaml:75]:
```suggestion
    requirements:
      host:
        - cuda-version {{ cuda_version }}
      run:
        - {{ pin_compatible("cuda-version", max_pin="x.x") }}
        - cuda-cupti >={{ version }}
```

**Review by bdice — COMMENTED** [review]
@jakirkham Good call on using `pin_compatible`. I reviewed this recipe before looking back at cublas, so I forgot that. It's fixed now and I deleted the outdated suggestions.

File: recipes/cuda-cupti/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+      run:
+        - {{ pin_compatible("cuda-version", max_pin="x.x") }}
+        - cuda-cupti >={{ version }}
+    files:
+      - lib/*.so                  # [linux]
+      - targets/*/lib/*.so        # [linux]
+      - include                   # [linux]
+      - targets/*/include         # [linux]
+      - samples
+      - doc
```
Comment (bdice) [inline recipes/cuda-cupti/meta.yaml:68]: Should this package ship 11 MB of docs and samples? Feels like that could be dropped with no loss. I don't think users look for docs / samples in conda environments, unless we think this is a crucial part of being a "full CTK."

File: recipes/cuda-cupti/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+requirements:
+  build:
+    - {{ compiler("c") }}
+    - {{ compiler("cxx") }}
+
+outputs:
+  - name: cuda-cupti
+    test:
+      commands:
+        - test -f $PREFIX/lib/libcupti.so.*                # [linux]
```
Comment (robertmaynard) [inline recipes/cuda-cupti/meta.yaml:35]: Needs to be updated to check for `libnvperf_target` and `libnvperf_host`

File: recipes/cuda-cupti/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+      - doc
+      - Library\include           # [win]
+      - Library\lib\*.lib         # [win]
+
+  - name: cuda-cupti-static
+    build:
+      skip: True                        # [not linux]
+    test:
+      commands:
+        - test -L $PREFIX/lib/libcupti_static.a  # [linux]
```
Comment (robertmaynard) [inline recipes/cuda-cupti/meta.yaml:77]: Needs to be updated to check for `libnvperf_target` and `libnvperf_host`

File: recipes/cuda-cupti/meta.yaml (hunk 4 above)
Comment (robertmaynard) [reply recipes/cuda-cupti/meta.yaml:68]: Offline choice was to ship the docs and samples in a `cuda-cupti-doc` package.

File: recipes/cuda-cupti/build.sh [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+for i in `ls`; do
+	[[ $i == "build_env_setup.sh" ]] && continue
+	[[ $i == "conda_build.sh" ]] && continue
+	[[ $i == "metadata_conda_debug.yaml" ]] && continue
+	if [[ $i == "lib" ]] || [[ $i == "include" ]]; then
+		mkdir -p ${PREFIX}/${targetsDir}
+		mkdir -p ${PREFIX}/$i
+		cp -rv $i ${PREFIX}/${targetsDir}
+		for j in `ls $i`; do
+			ln -s ../${targetsDir}/$i/$j ${PREFIX}/$i/$j
```
Comment (robertmaynard) [inline recipes/cuda-cupti/build.sh:19]:
```suggestion
			ln -s ${PREFIX}/${targetsDir}/$i/$j ${PREFIX}/$i/$j
```
match the style in libcublas

File: recipes/cuda-cupti/build.sh [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+		done
+	else
+		cp -rv $i ${PREFIX}
```
Comment (robertmaynard) [inline recipes/cuda-cupti/build.sh:22]:
```suggestion
		cp -rv $i ${PREFIX}
                # Put all other files in targetsDir
               mkdi r -p ${PREFIX}/${targetsDir}/${PKG_NAME}
              cp -rv $i ${PREFIX}/${targetsDir}/${PKG_NAME}		
```
Match libcublas package, and fix the issue where we might create extra directories by mistake

File: recipes/cuda-cupti/build.sh [hunk 9]
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
+
+# Install to conda style directories
+[[ -d lib64 ]] && mv lib64 lib
+
```
Comment (robertmaynard) [inline recipes/cuda-cupti/build.sh:5]:
```suggestion
mkdir -p ${PREFIX}/lib
```

**Review by robertmaynard — CHANGES_REQUESTED** [review]

File: recipes/cuda-cupti/meta.yaml [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+  sha256: b65a34a3715e3ef2b835bdd36c43b21c20800325be1eda1e84ad4d80ee837c92  # [win]
+
+build:
+  number: 0
+  skip: true  # [osx]
+
+outputs:
+  - name: cuda-cupti
+    files:
+      - lib/libcupti*.so.*                    # [linux]
```
Comment (bdice) [inline recipes/cuda-cupti/meta.yaml:33]: Do we not need to ship `libnvperf_host.so` and `libnvperf_target.so` in the `$PREFIX/lib/` folder? If so, please make sure they are tested (these are the files that were missing in the `nvidia` channel package that are needed by nvbench).

File: recipes/cuda-cupti/meta.yaml [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+        - {{ compiler("c") }}
+        - {{ compiler("cxx") }}
+        - arm-variant * {{ arm_variant_type }}  # [aarch64]
+      host:
+        - cuda-version {{ cuda_version }}
+      run:
+        - {{ pin_compatible("cuda-version", max_pin="x.x") }}
+        - {{ pin_subpackage("cuda-cupti", exact=True) }}
+      run_constrained:
+        - arm-variant * {{ arm_variant_type }}  # [aarch64]
```
Comment (bdice) [inline recipes/cuda-cupti/meta.yaml:85]:
```suggestion
        - arm-variant * {{ arm_variant_type }}  # [aarch64]
        - cuda-cupti-static >={{ version }}
```

File: recipes/cuda-cupti/meta.yaml [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+        - if not exist %LIBRARY_BIN%\cupti64_{{ version.split(".")[0] }}.dll exit 1              # [win]
+    about:
+      home: https://developer.nvidia.com/cuda-toolkit
+      license_file: LICENSE
+      license: LicenseRef-NVIDIA-End-User-License-Agreement
+      license_url: https://docs.nvidia.com/cuda/eula/index.html
+      summary: Provides libraries to enable third party tools using GPU profiling APIs.
+      description: |
+        Provides libraries to enable third party tools using GPU profiling APIs.
+      doc_url: https://docs.nvidia.com/cuda/index.html
```
Comment (bdice) [inline recipes/cuda-cupti/meta.yaml:62]: Consider using https://docs.nvidia.com/cuda/cupti/index.html as the `doc_url` throughout this file.

**Review by bdice — COMMENTED** [review]
Minor requests. Recipe looks fine aside from those. I haven't checked the package contents yet.

File: recipes/cuda-cupti/meta.yaml (hunk 10 above)
Comment (adibbley) [reply recipes/cuda-cupti/meta.yaml:33]: They are in `-dev` as they are unversioned only.

[Tests below](https://github.com/conda-forge/staged-recipes/pull/22355/files#diff-fdf8b09d413622d288ab00734d612c4ff1cf43048e12cac549314347a6605e9bR89-R90)

File: recipes/cuda-cupti/meta.yaml [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+  number: 0
+  skip: true  # [osx]
+
+test:
+  commands:
+    - test -L $PREFIX/lib/libcupti.so.{{ version.split(".")[0] }}                            # [linux]
+    - test -L $PREFIX/lib/libcupti.so.{{ lib_version }}                                      # [linux]
+    - test -L $PREFIX/targets/{{ target_name }}/lib/libcupti.so.{{ version.split(".")[0] }}  # [linux]
+    - test -f $PREFIX/targets/{{ target_name }}/lib/libcupti.so.{{ lib_version }}            # [linux]
+    - if not exist %LIBRARY_BIN%\cupti64_{{ version.split(".")[0] }}.dll exit 1              # [win]
```
Comment (jakirkham) [inline recipes/cuda-cupti/meta.yaml:37]: This test appears to be [failing on CI]( https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=688664&view=logs&j=240f1fee-52bc-5498-a14a-8361bde76ba0&t=2fb5b0a7-737c-5f5c-8f3f-f6db6174bacf&l=1670 )

Looking at the `redist`, it appears this is named differently. So guessing this test should be something like the change below. Though please let me know if I'm missing something

```suggestion
    - if not exist %LIBRARY_BIN%\cupti64_{{ lib_version }}.dll exit 1              # [win]
```

**Review by jakirkham — COMMENTED** [review]
Saw a test failure on CI. Proposed a (potential) fix below. Though please double check

File: recipes/cuda-cupti/meta.yaml [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+      license: LicenseRef-NVIDIA-End-User-License-Agreement
+      license_url: https://docs.nvidia.com/cuda/eula/index.html
+      summary: Provides libraries to enable third party tools using GPU profiling APIs.
+      description: |
+        Provides libraries to enable third party tools using GPU profiling APIs.
+      doc_url: https://docs.nvidia.com/cuda/cupti/index.html
+
+  - name: cuda-cupti-static
+    build:
+      skip: True                        # [not linux]
```
Comment (bdice) [inline recipes/cuda-cupti/meta.yaml:113]:
```suggestion
      skip: True  # [not linux]
```

**Review by bdice — APPROVED** [review]

File: recipes/cuda-cupti/meta.yaml [hunk 15]
```diff
@@ -0,0 +1,177 @@
...
+        - {{ pin_subpackage("cuda-cupti", exact=True) }}
+      run_constrained:
+        - arm-variant * {{ arm_variant_type }}  # [aarch64]
+        - cuda-cupti-static >={{ version }}
+    test:
+      commands:
+        - test -L $PREFIX/lib/libcupti.so                                    # [linux]
+        - test -L $PREFIX/lib/libnvperf_host.so                              # [linux]
+        - test -L $PREFIX/lib/libnvperf_target.so                            # [linux]
+        - test -f $PREFIX/targets/{{ target_name }}/include/cupti.h          # [linux]
```
Comment (kkraus14) [inline recipes/cuda-cupti/meta.yaml:95]: Should the header exist in `$PREFIX/include` as well via a symlink?

Comment (kkraus14) [reply recipes/cuda-cupti/meta.yaml:95]: See comment here: https://github.com/conda-forge/staged-recipes/pull/22127#discussion_r1168746777

**Review by kkraus14 — APPROVED** [review]

**Review by robertmaynard — APPROVED** [review]

**Comment by jakirkham** [comment]
Thanks all! 🙏

Let's follow up on anything else in the feedstock 🙂

**Pull request merged**